src/main/java/org/elasticsearch/script/ScriptService.java

```diff
@@ -217,60 +216,70 @@ private ScriptEngineService getScriptEngineServiceForFileExt(String fileExtensio
     /**
      * Checks if a script can be executed and compiles it if needed, or returns the previously compiled and cached script.
      */
-    public CompiledScript compile(String lang,  String script, ScriptType scriptType, ScriptContext scriptContext) {
-        assert script != null;
-        assert scriptType != null;
-        assert scriptContext != null;
+    public Script compile(Script script) {
+        assert script.key != null;
+        assert script.type != null;
+        assert script.context != null;
+
+        String lang = script.lang;
 
         if (lang == null) {
             lang = defaultLang;
         }
 
         ScriptEngineService scriptEngineService = getScriptEngineServiceForLang(lang);
-        if (canExecuteScript(lang, scriptEngineService, scriptType, scriptContext) == false) {
-            throw new ScriptException("scripts of type [" + scriptType + "], operation [" + scriptContext.getKey() + "] and lang [" + lang + "] are disabled");
+        if (canExecuteScript(lang, scriptEngineService, script.type, script.context) == false) {
+            throw new ScriptException("scripts of type [" + script.type + "], operation [" + script.context.getKey() + "] and lang [" + lang + "] are disabled");
         }
-        return compileInternal(lang, script, scriptType);
+
+        return compileInternal(script);
     }
 
     /**
      * Compiles a script straight-away, or returns the previously compiled and cached script, without checking if it can be executed based on settings.
      */
-    public CompiledScript compileInternal(String lang,  String script, ScriptType scriptType) {
-        assert script != null;
-        assert scriptType != null;
+    public Script compileInternal(Script script) {
```

Comment:
What's the advantage of having this return Script instead of CompiledScript? Maybe we would just wrap the whole Script around CompiledScript instead if needed?